simulated_beacon/simulated_beacon.go

```diff
@@ -0,0 +1,236 @@
+	} else {
+		popCount = len(w.pending)
+	}
+
+	popped := make([]*types.Withdrawal, popCount)
+	copy(popped[:], w.pending[0:popCount])
+	w.pending = append([]*types.Withdrawal{}, w.pending[popCount:]...)
+	return popped
+}
+
+// add adds a withdrawal to the queue
+func (w *withdrawals) add(withdrawal *types.Withdrawal) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if withdrawal.Index < w.next {
+		return fmt.Errorf("withdrawal has index (%d) less than or equal to latest received withdrawal index (%d)", withdrawal.Index, w.next-1)
+	}
+	w.next = withdrawal.Index + 1
+	w.pending = append(w.pending, withdrawal)
+	return nil
+}
+
+type SimulatedBeacon struct {
+	ctx          context.Context
+	cancel       context.CancelFunc
+	eth          *eth.Ethereum
+	period       time.Duration
+	withdrawals  withdrawals
+	feeRecipient common.Address
+	// mu gates concurrent access to the feeRecipient
+	mu sync.Mutex
+}
+
+func NewSimulatedBeacon(eth *eth.Ethereum) *SimulatedBeacon {
+	chainConfig := eth.APIBackend.ChainConfig()
+	if chainConfig.Dev == nil {
+		log.Crit("incompatible pre-existing chain configuration")
+	}
+
+	return &SimulatedBeacon{
+		eth:          eth,
+		period:       time.Duration(chainConfig.Dev.Period) * time.Second,
+		withdrawals:  withdrawals{[]*types.Withdrawal{}, 0, sync.Mutex{}},
+		feeRecipient: common.Address{},
+	}
+}
+
+func (c *SimulatedBeacon) setFeeRecipient(feeRecipient *common.Address) {
+	c.mu.Lock()
+	c.feeRecipient = *feeRecipient
+	c.mu.Unlock()
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.loop()
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service
+func (c *SimulatedBeacon) Stop() error {
+	c.cancel()
+	return nil
+}
+
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *SimulatedBeacon) loop() {
+	var (
+		ticker             = time.NewTicker(time.Millisecond * 100)
+		lastBlockTime      = time.Now()
+		engineAPI          = catalyst.NewConsensusAPI(c.eth)
+		header             = c.eth.BlockChain().CurrentHeader()
+		curForkchoiceState = engine.ForkchoiceStateV1{
+			HeadBlockHash:      header.Hash(),
+			SafeBlockHash:      header.Hash(),
+			FinalizedBlockHash: header.Hash(),
+		}
+	)
+
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.BitLen() == 0 {
+		if _, err := engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, nil); err != nil {
+			log.Crit("failed to initiate PoS transition for genesis via Forkchoiceupdated", "err", err)
+		}
+	}
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			break
+		case curTime := <-ticker.C:
+			if curTime.Unix() > lastBlockTime.Add(c.period).Unix() {
+				c.mu.Lock()
+				feeRecipient := c.feeRecipient
+				pendingWithdrawals := c.withdrawals.pop()
+				c.mu.Unlock()
```

Comment:
Also, maybe you could simplify the inner for-loop (and make some of those block-building messages go away, if you do the check on pending state earlier. Like this:

```golang 
		case curTime := <-ticker.C:
			if curTime.Unix() <= lastBlockTime.Add(c.period).Unix() {
				continue
			}
			c.mu.Lock()
			feeRecipient := c.feeRecipient
			pendingWithdrawals := c.withdrawals.pop()
			c.mu.Unlock()

			pendingHeader, err := c.eth.APIBackend.HeaderByNumber(context.Background(), rpc.PendingBlockNumber)
			if err != nil {
				log.Crit("failed to get pending block header", "err", err)
			}
			// don't build a block if we don't have pending txs or withdrawals
			if pendingHeader.TxHash == types.EmptyTxsHash && len(pendingWithdrawals) == 0 {
				continue
			}
			// Looks like it's time to build us a block!
```